pkg/httphelper/httphelper_test.go

```diff
@@ -0,0 +1,206 @@
+package httphelper
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus/testutil"
+)
+
+func TestWriteHeader(t *testing.T) {
+	testcases := []struct {
+		name       string
+		statusCode int
+	}{
+		{
+			"StatusOK",
+			http.StatusOK,
+		},
+		{
+			"StatusNotFound",
+			http.StatusNotFound,
+		},
+	}
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			rr := httptest.NewRecorder()
+			trw := &TraceResponseWriter{ResponseWriter: rr, statusCode: http.StatusOK}
+			trw.WriteHeader(tc.statusCode)
+			if rr.Code != tc.statusCode {
+				t.Errorf("mismatch between expected and actual response headers: expected %s, got %s", http.StatusText(tc.statusCode), http.StatusText(rr.Code))
+			}
+			if trw.statusCode != tc.statusCode {
+				t.Errorf("mismatch between expected and actual TraceResponseWriter headers: expected %s, got %s", http.StatusText(tc.statusCode), http.StatusText(trw.statusCode))
+			}
+		})
+
+	}
+
+}
+
+func TestWrite(t *testing.T) {
+	testcases := []struct {
+		name         string
+		responseBody string
+	}{
+		{
+			"SimpleText",
+			"Simple text to test respone size",
+		},
+	}
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			rr := httptest.NewRecorder()
+			trw := &TraceResponseWriter{ResponseWriter: rr, statusCode: http.StatusOK}
+			resp := []byte(tc.responseBody)
+			_, err := trw.Write(resp)
+			if err != nil {
+				t.Fatalf("failed to write to TraceResponseWriter")
+			}
+			if rr.Body.String() != tc.responseBody {
+				t.Errorf("mismatch between expected and actual response body: expected %s, got %s", tc.responseBody, rr.Body.String())
+			}
+			if trw.size != len(resp) {
+				t.Errorf("mismatch between expected and actual TraceResponseWriter size: expected %d, got %d", len(resp), trw.size)
+			}
+		})
+	}
+}
+
+var metrics = NewMetrics("testnamespace")
+
+var handleMetricsCalls = 0
+var recordErrorCalls = 0
```

Comment:
Huh do these *need* to be globals? The `metrics` seems to be shared between the two following tests and the globals are incremented in individual testcases (which is not thread-safe which will surprise anyone who will try to parallelize these tests in the future). Is there a way how to create these separately for each testcase? If yes then we should use it, if not, then this deserves an explanatory paragraph about why these are globals shared from multiple places.

Reply:
I think they need to be globals, initially I was initializing a metrics variable for each test but I ended up getting an error saying that the same metrics cant be registered multiple times which is why I was forced to make them globals. The call counters were needed to keep track of the number of times the tests were called. I guess I can use closures(?) for the call counters. wdyt?

Reply:
The counter does definitely not need to be a global and should be testcase-local. The metrics currently need to be (binary) global because `NewMetrics` calls `prometheus.MustRegister`. I suggest to add a new constructor that is used in `NewMetrics` for only constructing the metrics without registering  them to allow to also have the metrics test-case local (`testutil.CollectAndCompare` does not require the metric to be registered)

Reply:
So I would be making a new constructor only to allow for testing?